fix: tests without deprecation warnings

Lewiscowles1986 opened the pull request:
- Updates packages
- Removes branching logic about composer 1.x
- Ensures a Promise is returned when downloading file
- Documented php-cs-fixer dependency as a dev dependency
- fixed GitHub actions build matrix for php 8.1 - 8.3

Lewiscowles1986 committed:
fix: tests without deprecation warnings

- Updates packages
- Removes branching logic about composer 1.x
- Ensures a Promise is returned when downloading file

Lewiscowles1986 committed:
fix: php 8.1 compatibility

Lewiscowles1986 committed:
fix: php-cs-fixer documented as dev dependency

Lewiscowles1986 committed:
fix: build matrix to be 8.1, 8.2, 8.3

tienvx commented on composer.json:
```diff
@@ -18,17 +18,18 @@
         "leongrdic/smplang": "^1.0.2",
-        "symfony/filesystem": "^5.4 || ^6.0",
-        "symfony/finder": "^4.4.23 || ^5.4 || ^6.0"
+        "symfony/filesystem": "^6.0 || ^7.0",
+        "symfony/finder": "^6.0 || ^7.0"
```
I think Symfony 5.4 is still supported. So we still want to support it https://symfony.com/releases

Lewiscowles1986 replied:
<img width="825" alt="Screenshot 2024-03-16 at 17 01 33" src="https://github.com/pact-foundation/composer-downloads-plugin/assets/2605791/95163dab-a71a-4524-bc9a-373698535b1c">

While 5.4 may well still be supported, PACT is now 8.1 only, so this is all it needs. I did have some issues when just using 7.0 and when trying to support all 3. These two do not exhibit the same problems.

tienvx replied:
I wonder what are those issues. I add Symfony 5.4 back and test and look like it works (apart of Windows's rar issue) https://github.com/pact-foundation/composer-downloads-plugin/actions/runs/8309331776/job/22740567129

tienvx replied:
I suggest this change:

```diff
+         "symfony/finder": "^5.4 || ^6.4 || ^7.0"
```

tienvx replied:
> PACT is now 8.1 only, so this is all it needs

But PACT list `tienvx/composer-downloads-plugin` (soon will be `pact-foundation/composer-downloads-plugin`) in `require`, not `require-dev`, and I'm worrying that some projects that are using Symfony 5.4 will not able to install PACT because of dependency conflicts (e.g. those project only require `"symfony/finder: ^5.4"`)

tienvx commented on composer.json:
```diff
@@ -18,17 +18,18 @@
         "phpunit/phpunit": "^9.5.10",
-        "symfony/process": "^4.4 || ^5.4 || ^6.0",
-        "php-vfs/php-vfs": "^1.4"
+        "symfony/process": "^6.0 || ^7.0",
```
Same comment about Symfony 5.4

Lewiscowles1986 replied:
same response for all of Symfony

tienvx commented on src/Handler/FileHandler.php:
```diff
@@ -42,28 +42,17 @@ protected function download(Composer $composer, IOInterface $io): void
+        $promise->then(static function ($res) use (&$file) {
+            $file = $res;
+
+            return \React\Promise\resolve($res);
```
Do you know which version of composer require this change?

Do we need to update version of `composer-plugin-api` or `composer/composer` to that version?

tienvx replied:
I think we already test after running `composer install --prefer-lowest` and it work fine, so no need to update those packages.

Let me check why I miss this before.

Lewiscowles1986 replied:
I believe it might have been weaker type-checking causing the issue in PHP core. But I will look into this.

tienvx replied:
I tried to remove this line and the tests still passed (lowest dependency). So it's not required as I assume. I will resolve as it's a welcome change.

tienvx replied:
I see the problem now on highest dependency:

```
There were 2 errors:

1) LastCall\DownloadsPlugin\Tests\Unit\FileHandlerTest::testInstall
TypeError: Mock_PromiseInterface_5f24811c::then(): Return value must be of type React\Promise\PromiseInterface, null returned

/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/FileHandler.php:47
/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/BaseHandler.php:65
/home/tien/Projects/pact/composer-downloads-plugin/tests/Unit/Handler/BaseHandlerTestCase.php:91

2) LastCall\DownloadsPlugin\Tests\Unit\PharHandlerTest::testInstall
TypeError: Mock_PromiseInterface_5f24811c::then(): Return value must be of type React\Promise\PromiseInterface, null returned

/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/FileHandler.php:47
/home/tien/Projects/pact/composer-downloads-plugin/src/Handler/BaseHandler.php:65
/home/tien/Projects/pact/composer-downloads-plugin/tests/Unit/Handler/BaseHandlerTestCase.php:91


```

So this change + upgrade to composer 2.3 LGTM

Lewiscowles1986 replied:
https://github.com/composer/composer/pull/10547/files#diff-b99c0f08ed0d8f48cb9926a9f015a5cc1fadc6a5c4111c106d788c64264bcb30L187

Which PHP version were you running under? You can see in prior commits it failing based upon this.

Lewiscowles1986 replied:
This is php 8.1

<img width="955" alt="Screenshot 2024-03-16 at 18 25 14" src="https://github.com/pact-foundation/composer-downloads-plugin/assets/2605791/ed8d972e-6501-43b5-8e41-86dcc30dfb73">

I'm sure it's the composer version that we're running into in it ever passing. It should not past 8.1 and compoer 2.3

Lewiscowles1986 committed:
docs: fix dependency floor at composer 2.3

https://github.com/composer/composer/pull/10547/files#diff-b99c0f08ed0d8f48cb9926a9f015a5cc1fadc6a5c4111c106d788c64264bcb30L187 seems to be where promise breakage comes from

Lewiscowles1986 committed:
chore: restore old version of symfony - review feedback

tienvx committed:
deps: Upgrade Symfony 6.0 to 6.4 (supported version)

tienvx commented:
Thanks @Lewiscowles1986 for helping me on this
